### DISCUSSION TITLE
[08S01] [FreeTDS][SQL Server]Bad token from the server: Datastream processing out of sync (20020) (SQLExecDirectW)

### DISCUSSION THREAD

**Discussion by vishakraj25** [sqlalchemy/sqlalchemy#9137] [Usage Questions]
Hello,

I am using the MSSql server docker - 
    
    1> select @@version
    2> go
    Microsoft SQL Server 2017 (RTM-CU22) (KB4577467) - 14.0.3356.20 (X64) 
	      Aug 20 2020 22:33:27
	      Copyright (C) 2017 Microsoft Corporation
	      Express Edition (64-bit) on Linux (Ubuntu 16.04.7 LTS)                                                                                                             

In other docker container, I am running the flask server with nginix 
versions :
Python=3.6.15
Flask=1.0.2
SQLAlchemy=1.3.8

these two docker containers are running on the ubuntu os 20.04.4 LTS in the same server

I used to connect the database running on the sql server in flask with mssql+pyodbc

	engine = create_engine(
	    'mssql+pyodbc:///?odbc_connect=%s' % (
	        urllib.parse.quote_plus(
	            'DRIVER={/usr/lib/x86_64-linux-gnu/odbc/libtdsodbc.so};SERVER=localhost;'
	            'DATABASE=alpha_info;UID=sa;PWD=admin@21;port=1433;'
	            'TDS_Version=8.0;')), isolation_level='READ COMMITTED')

I developed many APIs, which are all working so far good

For every request, I created the separate scoped_session and do the queries and finally, I removed the scoped_session, like this 

    def __init__(self):
        self.Session = sessionmaker(bind=engine)

    def get_result(self):
    	result = []
    	try:
	        Session = scoped_session(self.Session)
	        session = Session()
	        result = session.query(Table1).all()
        except Exception as e:
        	logging.exception(f'Failed to get details-{e}')
        finally:
	        session.close()
	        Session.remove()
	        return result

When opened the main page in the browser, there are three API calling one after another, lets say API1, API2, API3 - all APIs are just querying the database using `select` and return the data like the `get_result` API.

The issue:
When ever I refreshed the main page, these three APIs are called. In that, I got the result for any two of the API and I did not getting the result of the other API.

When I check the logs, I got this

> 2023-01-22 19:07:50,202 - ERROR - sqlalchemy.pool.impl.QueuePool - Exception during reset or similar
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 717, in __getattr__
    return getattr(self.comparator, key)
AttributeError: 'Comparator' object has no attribute '__clause_element__'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool/base.py", line 680, in _finalize_fairy
    fairy._reset(pool)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/pool/base.py", line 867, in _reset
    pool._dialect.do_rollback(self)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 502, in do_rollback
    dbapi_connection.rollback()
pyodbc.Error: ('HY000', 'The driver did not supply an error!')
2023-01-22 19:07:50,203 - ERROR - root - Exception in get_traffic_result, (pyodbc.OperationalError) ('08S01', '[08S01] [FreeTDS][SQL Server]Bad token from the server: Datastream processing out of sync (20020) (SQLExecDirectW)')
(Background on this error at: http://sqlalche.me/e/e3q8)
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
pyodbc.OperationalError: ('08S01', '[08S01] [FreeTDS][SQL Server]Bad token from the server: Datastream processing out of sync (20020) (SQLExecDirectW)')
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "./config.py", line 6178, in get_traffic_result
    .filter(Table1.status==1).count()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3578, in count
    return self.from_self(col).scalar()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3322, in scalar
    ret = self.one()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3292, in one
    ret = self.one_or_none()
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3261, in one_or_none
    ret = list(self)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3334, in __iter__
    return self._execute_and_instances(context)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3359, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (pyodbc.OperationalError) ('08S01', '[08S01] [FreeTDS][SQL Server]Bad token from the server: Datastream processing out of sync (20020) (SQLExecDirectW)')
(Background on this error at: http://sqlalche.me/e/e3q8)

This kind of error is occurred in any one of the three API,

In some other time, I got the error in one of the API, and in the same call to API after the error throws, I got the result, 

Lets, take this log event -

> 2023-01-22 20:10:16,625 - ERROR - root - Exception in get_status, (pyodbc.OperationalError) ('08S01', '[08S01] [FreeTDS][SQL Server]Bad token from the server: Datastream processing out of sync (20020) (SQLExecDirectW)')
(Background on this error at: http://sqlalche.me/e/e3q8)
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
pyodbc.OperationalError: ('08S01', '[08S01] [FreeTDS][SQL Server]Bad token from the server: Datastream processing out of sync (20020) (SQLExecDirectW)')
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "./config.py", line 6276, in get_status
    Table1.type==2
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3178, in all
    return list(self)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3334, in __iter__
    return self._execute_and_instances(context)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/orm/query.py", line 3359, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (pyodbc.OperationalError) ('08S01', '[08S01] [FreeTDS][SQL Server]Bad token from the server: Datastream processing out of sync (20020) (SQLExecDirectW)')
(Background on this error at: http://sqlalche.me/e/e3q8)
2023-01-22 20:10:16,628 - INFO - root - get_status started **###Here the same api throws the above error, shown as started**
2023-01-22 20:10:17,933 - INFO - root - get_status completed **###Here the api completed successfully**

**Note:** I printed the 'get_status started' when API3 start and 'get_status completed' when API3 successfully executed
From the above log, at the time of `2023-01-22 20:10:16,625`, I got the error
and at the time of `2023-01-22 20:10:16,628`, I got the API3 executed successfully in the same call from the main page - How is it working, does the sqlalchemy runs the API one more time, if there is any error happened?

I tried added the timeout in the create_engine - 

	engine = create_engine(
	    'mssql+pyodbc:///?odbc_connect=%s' % (
	        urllib.parse.quote_plus(
	            'DRIVER={/usr/lib/x86_64-linux-gnu/odbc/libtdsodbc.so};SERVER=localhost;'
	            'DATABASE=alpha_info;UID=sa;PWD=admin@21;port=1433;'
	            'TDS_Version=8.0;')), isolation_level='READ COMMITTED', connect_args={'options':'-c lock_timeout=3000 -c statement_timeout=3000', 'timeout':40}, max_overflow=10, pool_size=30, pool_timeout=60)

but, I got the same error. 

When I run all the API separately, I got the result without any issue

Is these kind of error happen, when the API calls the database at the same time, or
when one API trying to connect the database when other API currently executes the query in the database?

From the sqlalchemy doc, I get to know, that the scoped_session create threads form the sessionmaker for each request. I cross checked this by checking the `id(session)` in every API request - which is correct, the scoped_session creates new thread for every request

How to solve this issue?

Thanks

**Comment by zzzeek** [comment]
if this were on the Python side it might have to do with a concurrency issue such as non-threadsafe behavior in the application, like sharing a Session, a Connection, or objects that are attached to a Session in more than one thread at a time.

at the same time I see you're using freetds for connectivity and I would strongly recommend using Microsoft's ODBC drivers with pyodbc; freetds' drivers are much less predictable.

**Reply by zzzeek** [reply]
I'd also add this application has no reason to use scoped_session() as illustrated:

```py
def __init__(self):
    self.Session = sessionmaker(bind=engine)

def get_result(self):
	result = []
	try:
        Session = scoped_session(self.Session)
        session = Session()
        result = session.query(Table1).all()
    except Exception as e:
    	logging.exception(f'Failed to get details-{e}')
    finally:
        session.close()
        Session.remove()
        return result
```

scoped_session is only useful if it's a globally scoped object that many parts of an application all access globally.    There's no reason to create one inside of a local scope and dispose it.    your code could be:

```py
def __init__(self):
    self.Session = sessionmaker(bind=engine)

def get_result(self):
    result = []
    try:
         session = Session()
         result = session.query(Table1).all()
    except Exception as e:
        session.rollback()
    	logging.exception(f'Failed to get details-{e}')
    finally:
        session.close()
        return result
```

note also session.rollback() when an error is caught, not strictly necessary here since you are calling .close() each time.

**Reply by vishakraj25** [reply]
@zzzeek, At first, I initialized the scoped_session globally in `__init__` function, I got the same error, then in a blog, I read that for each request it is good to have separate session for each API, so, I modified the code to have the scoped_session for each API but the error stays the same

In the real time, the different API can be called at same time, for that the scoped_session will create different thread to query the database even if the same API called from different end - I tested this scenario with multiprocessing python

And, I dont think this issue is from python side, because the APIs are not sharing the connections/sesssion variables from one to another but the APIs are sharing the output data only

The error clearly stated this from database side, when executing the `select` query. 

The thing is - it is hard to track, beccause every time the error occurring in different api

**Reply by CaselIT** [reply]
> In the real time, the different API can be called at same time, for that the scoped_session will create different thread to query the database even if the same API called from different end - I tested this scenario with multiprocessing python

scoped session creates no thread at all. I suggestion you to re-check the documentation https://docs.sqlalchemy.org/en/14/orm/contextual.html
also as mike mentioned, if you create a session inside a method, use it in the method and then close it at the end, there is no need to use a scoped session even when multiple threads are used

**Reply by vishakraj25** [reply]
@CaselIT, I go through the doc, I modified the code like this -

		def __init__(self):
			session = sessionmaker(bind=engine)
			self.Session = scoped_session(session)

		def get_result(self):
			result = []
			try:
				session = self.Session()
				result = session.query(Table1).all()
			except Exception as e:
				logging.exception(f'Failed to get details-{e}')
			finally:
				session.close()
				return result


After changing the code, I still got the same error

**Reply by vishakraj25** [reply]
Sometimes, I got this error - 

> sqlalchemy.exc.NoSuchColumnError: "Could not locate column in row for column 'Table1.type'"

How the sqlalchemy does not detect the column when that table is already using from other APIs?

or this error

> sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically.

It would be great relief if I get any solution, thanks

**Reply by vishakraj25** [reply]
@zzzeek ,
> at the same time I see you're using freetds for connectivity and I would strongly recommend using Microsoft's ODBC drivers with pyodbc; freetds' drivers are much less predictable.

Could you please share some instruction to use Microsoft's ODBC drivers with pyodbc, instead of using freetds?

In create_engine, the driver is from odbc only - `/usr/lib/x86_64-linux-gnu/odbc/libtdsodbc.so`, then the TDS_Version parameter is not needed?

**Reply by zzzeek** [reply]
> @zzzeek ,
> 
> > at the same time I see you're using freetds for connectivity and I would strongly recommend using Microsoft's ODBC drivers with pyodbc; freetds' drivers are much less predictable.
> 
> Could you please share some instruction to use Microsoft's ODBC drivers with pyodbc, instead of using freetds?
> 
> In create_engine, the driver is from odbc only - `/usr/lib/x86_64-linux-gnu/odbc/libtdsodbc.so`, then the TDS_Version parameter is not needed?

URLs for the MSFT drivers most simply look like this:

    engine = create_engine("mssql+pyodbc://scott:tiger@myhost:port/databasename?driver=ODBC+Driver+17+for+SQL+Server")

that is using a direct host connection without having to configure an ODBC DSN.

The change in driver may or may not alleviate the issue but you may very well get different behavior that will perhaps shed some light on the source of the issue.

**Reply by zzzeek** [reply]
> Sometimes, I got this error -
> 
> > sqlalchemy.exc.NoSuchColumnError: "Could not locate column in row for column 'Table1.type'"
> 
> How the sqlalchemy does not detect the column when that table is already using from other APIs?

it would greatly depend on if the code that causes this error (seen in the stacktrace) has this issue every time, in which case something is wrong with the query construction (also SQLAlchemy prior to 1.4 was a little more prone to this issue overall), or if it's sporadic, which suggests concurrency issues where the state of the connection is being corrupted.


> 
> or this error
> 
> > sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically.

this strongly suggests the connection is in unexpected states (usually due to concurrency issues), where like the previous error, a query was executed but the connection's state changes before results are fetched.


> 
> It would be great relief if I get any solution, thanks

you need to identify both the exact circumstances which cause this issue as well as it would be helpful to identify what event precipitated the code having this problem, as well as if it is environmentally specific:

1. can you reproduce reliably?    like create a script that will produce the error every time?
2. can it be reproduced in a development environment from a clean installation?
3. does it depend on certain kind of data being present in the database
5. is there a version of the code at which the issue begins to be reproducible (e.g. use `git bisect`)
6. can the components of the reproduction case be reduced?   when the reproduction case is isolated, being removing components from the application's functionality; remove DB calls and other components to see at which point the error stops happening

**Reply by vishakraj25** [reply]
@zzzeek, thanks for writting

In creating engine, I added the `timeout` parameter as `40`. 

I still got the issue, then after one/two days later only, I started working on this issue again, from that time I didnt face this issue, every thing is working properly

But, One **Event** to note here is, that if we call the multiple APIs continuously one after another - the first 4-5 transaction to API, it responds fastly as expected, but for the next two-three API call, it takes approx. 40 sec to load - is this because of the `timeout` parameter.  Then, I checked the start and finish time of the each API, all the APIs are taking only 3-5 sec

If I restart the docker containers, the issue happening again, and after some time, the issue is not occurring, and the **Event** started

could you please help me to understand this flow

thanks
